README.md
# Plug-Range

An [Elixir Plug](http://github.com/elixir-lang/plug) to add HTTP Range Requests

## Usage

1. Add this plug to your `mix.exs` dependencies:

```elixir
def deps do
  # ...
  {:plug_range, git: "https://github.com/TheSquad/Plug-Range.git"}
  #...
end
```

When used with phoenix framework, please note that it should be called before Plug.Static
otherwise most assets will be delivered by it before Plug.Range had a chance.

```elixir
defmodule YourApp.Endpoint do
  use Phoenix.Enpoint, otp_app: :your_app

  plug PlugRange

  plug Plug.Static,
  at: "/", from: :my_project, gzip: false,
  only: ~w(css fonts images js favicon.ico robots.txt)

  # ...

  plug YourApp.Router
end
```

## Serving HTTP Range Requests

Modern web browser sometime uses Range Requests to get mostly static files, like images, videos, etc...

For example : when using the ```<video></video>``` tag on your website, not serving HTTP Range Requests
will make Safari fail to retreive the video to play it.

As soon as the request is a "range", the Plug will take over and treat it as such. if not, it will return
the conn like nothing happened.

## RFC 7233

For now PlugRange is not RFC compliant, but I will try to make it compliant as soon as possible.
You are welcome to help in this purpose


CHANGELOG.md
Removing warnings in Elixir 1.4


test/plug_range_test.exs
defmodule PlugRangeTest do
  use ExUnit.Case
  use Plug.Test
  doctest PlugRange

  test "the truth" do
    assert 1 + 1 == 2
  end

  @opts PlugRange.init([])

  test "no range request" do
    conn = conn(:get, "/lorem.txt")
    conn_changed? = PlugRange.call(conn, @opts)
    assert conn == conn_changed?
  end

  test "range request 0-" do
    conn = conn(:get, "/lorem.txt")
    |> Plug.Conn.put_req_header("range", "bytes=0-")

    conn_changed? = PlugRange.call(conn, @opts)
    assert conn == conn_changed?
  end

  test "range request 0-32" do
    conn = conn(:get, "/lorem.txt")
    |> Plug.Conn.put_req_header("range", "bytes=0-32")

    conn = PlugRange.call conn, @opts

    assert conn.state == :sent
    assert conn.status == 206
    res_string = File.stream!("priv/static/lorem.txt", [], 33) |> Stream.take(1) |> Enum.to_list |> to_string
    assert conn.resp_body == res_string
  end

  test "range request -32" do
    conn = conn(:get, "/lorem.txt")
    |> Plug.Conn.put_req_header("range", "bytes=-32")

    conn = PlugRange.call conn, @opts

    assert conn.state == :sent
    assert conn.status == 206
    fs = File.stat! "priv/static/lorem.txt"
    res_string = File.read!("priv/static/lorem.txt") |> String.slice(fs.size - 32, 32)

    assert conn.resp_body == res_string
  end

  test "bad range request" do
    conn = conn(:get, "/lorem.txt")
    |> Plug.Conn.put_req_header("range", "bytes=-")

    conn = PlugRange.call conn, @opts
    assert conn.state == :sent
    assert conn.status == 416
  end

  test "bad range request 2" do
    conn = conn(:get, "/lorem.txt")
    |> Plug.Conn.put_req_header("range", "bytes=toto")

    conn = PlugRange.call conn, @opts
    assert conn.state == :sent
    assert conn.status == 416
  end

  test "bad range request 3" do
    conn = conn(:get, "/lorem.txt")
    |> Plug.Conn.put_req_header("range", "by")

    conn = PlugRange.call conn, @opts
    assert conn.state == :sent
    assert conn.status == 416
  end

end


mix.exs
defmodule PlugRange.Mixfile do
  use Mix.Project

  def project do
    [app: :plug_range,
     version: "0.0.3",
     elixir: "~> 1.2",
     description: description(),
     package: package(),
     build_embedded: Mix.env == :prod,
     start_permanent: Mix.env == :prod,
     deps: deps()]
  end

  # Configuration for the OTP application
  #
  # Type "mix help compile.app" for more information
  def application do
    [applications: [:logger]]
  end

  # Dependencies can be Hex packages:
  #
  #   {:mydep, "~> 0.3.0"}
  #
  # Or git/path repositories:
  #
  #   {:mydep, git: "https://github.com/elixir-lang/mydep.git", tag: "0.1.0"}
  #
  # Type "mix help deps" for more examples and options

  defp deps do
    [
      # {:cowboy, "~> 1.0.0"},
      {:plug, ">= 0.0.0", only: :test, optional: true},
      {:ex_doc, ">= 0.0.0", only: :dev}
    ]
  end

  defp description do
  """
  An elixir plug that serves HTTP range requests
  """
  end

  defp package do
    [
      files: ~w(lib mix.exs README.md LICENSE CHANGELOG.md),
      maintainers: ["Morgan Segalis"],
      licenses: ["MIT"],
      links: %{
        "Github" => "http://github.com/TheSquad/plug_range",
      }
    ]
  end
end


config/config.exs
# This file is responsible for configuring your application
# and its dependencies with the aid of the Mix.Config module.
use Mix.Config

# This configuration is loaded before any dependency and is restricted
# to this project. If another project depends on this project, this
# file won't be loaded nor affect the parent project. For this reason,
# if you want to provide default values for your application for
# 3rd-party users, it should be done in your "mix.exs" file.

# You can configure for your application as:
#
#     config :plug_range, key: :value
#
# And access this configuration in your application as:
#
#     Application.get_env(:plug_range, :key)
#
# Or configure a 3rd-party app:
#
#     config :logger, level: :info
#

# It is also possible to import configuration files, relative to this
# directory. For example, you can emulate configuration per environment
# by uncommenting the line below and defining dev.exs, test.exs and such.
# Configuration from the imported file will override the ones defined
# here (which is why it is important to import them last).
#
#     import_config "#{Mix.env}.exs"


lib/plug_range.ex
defmodule PlugRange do
  @behaviour Plug
  # @allowed_methods ~w(GET HEAD)

  def init(options) do
    options
  end

  def call(conn, _opts) do
    if (Enum.empty?(Plug.Conn.get_req_header(conn, "range"))) do
      conn
    else
      file_path = "priv/static" <> conn.request_path
      if File.exists? file_path do

        stats = File.stat! file_path
        filesize = stats.size

        req = Regex.run(~r/bytes=([0-9]+)?-([0-9]+)?/, conn |> Plug.Conn.get_req_header("range") |> List.first)

        case req do
          p when p == nil or ["bytes=", "", "-"] == p or length(p) < 2 ->
            send_bad_range conn
          _ ->
            {req_start, _} =
            if (req |> Enum.at(1) == "") do
              {-1, :end_file}
            else
              req |> Enum.at(1)  |> Integer.parse
            end
            {req_end, _} = req |> Enum.at(2, filesize |> to_string) |> Integer.parse

            if (req_start == 0 && req_end == filesize) do
              conn
            else

              {req_start, length} = case req_start do
                                      -1 -> {filesize - req_end, req_end}
                                      _ -> {req_start, req_end - req_start + 1}
                                    end

              # file_end = ( filesize - 2) |> to_string

              conn
              |> Plug.Conn.put_resp_header("accept-ranges", "bytes")
              |> Plug.Conn.put_resp_header("content-range", "bytes #{req_start}-#{req_end}/#{filesize}")
              |> Plug.Conn.send_file(206, file_path, req_start, length)
              |> Plug.Conn.halt
            end
        end
      end
    end
  end

  def send_bad_range(conn) do
    conn
    |> Plug.Conn.put_resp_content_type("text/html")
    |> Plug.Conn.put_resp_header("content-length", bad_range() |> String.length |> to_string)
    |> Plug.Conn.send_resp(416, bad_range())
    |> Plug.Conn.halt
  end

  def bad_range do
    """
    <html>
    <head><title>416 Requested Range Not Satisfiable</title></head>
    <body bgcolor="white">
    <center><h1>416 Requested Range Not Satisfiable</h1></center>
    <hr><center>cowboy</center>
    </body>
    </html>
    """
  end
end
